### PR TITLE
Fix AlchemicalNonequilibriumLangevinIntegrator.reset()

### LINKED ISSUE #273: Parameters reset on reinitialization
I've been trying to do the following:
1.Take i number of steps with `openmmtools. AlchemicalNonequilibriumLangevinIntegrator`.
2. After i steps, set the masses of a subset of the system to zero.
3. Call `context.reinitialize()` so the changes to the system's masses are actually applied to the simulation.
4. Pass in the saved state before reinitialize so after reinitialization the context is the same as before (except the masses).

However when I coded this up I got behavior I didn't expect, namely that after reinitializing the `lambda_sterics` and `lambda_electrostatics` parameters were reset to `1.0`.

Some example print output related to the parameters
```python
parameters ['lambda_electrostatics', 'lambda_sterics']
parameters before reinitialization [0.0, 0.6666666666666667]
parameters after reinitialization [0.95, 1.0] (and after 1 step)
parameters after nsteps_neq steps [0.0, 0.6666666666666665]
```

I'm using a symmetric protocol for the `alchemical_functions`, so after nsteps_neq the electrostatics and sterics should both be `1.0` which they're not.

I found a strange fix to this behavior, which was to get an integrator variable (ex. `integrator.getGlobalVariableByName('lambda')` after every step taken. However strange thing was that if this was done immediately after reinitialization I got a segmentation fault.

```python
#after adding the above fix
parameters ['lambda_electrostatics', 'lambda_sterics']
parameters before reinitialization [0.0, 0.6666666666666667]
parameters after reinitialization [0.0, 0.6333333333333334] (and after 1 step)
parameters after nsteps_neq steps [0.9999999999999998, 1.0]
```

It's not clear to me whether this is an integrator bug, an OpenMM bug, or intended behavior from OpenMM I just wasn't expecting. Since you folks are more familiar with the `AlchemicalNonequilibriumLangevinIntegrator` I thought I'd ask here first for thoughts on this.

I've incuded example scripts to replicate the above issue in [param_issue.zip](https://github.com/choderalab/openmmtools/files/1223873/param_issue.zip), which contains:

`issue.py`, reproduces the issue with the `lambda_sterics` and `lambda_electrostatics` being reset
`fix.py`, is essentially the same code, with the `integrator.getGlobalVariableByName('lambda')` fix above to resolve the problem
`seg_fault.py`, the same code again, but with a `integrator.getGlobalVariableByName('lambda')` line immediately after reinitializing.

### PR DESCRIPTION
**Pull request by jchodera**
Due to the issues @sgill2 was noting in https://github.com/choderalab/openmmtools/issues/273, I've been looking into the implementation of `AlchemicalNonequilibriumLangevinIntegrator.reset()`. 

The previous implementation would reset `step`, which would cause the *next* call to `integrator.step()` to reset the statistics and context parameters at the beginning of the next step, but wouldn't actually reset everything as soon as `reset()` was called which led to surprising behavior. I've changed this to take a small step "false" step (using step = -1) to reset everything as soon as `.reset()` is called. This may slow things down a bit for integration, but may be much safer.

I also noticed we weren't actually testing `reset()`, so I refactored `tests/test_integrators.py:run_alchemical_langevin_integrator()` to use a single `Context` with a `CompoundIntegrator` that has both equilibrium and nonequilibrium integrators that uses `integrator.reset()` to reset statistics.

I'm not sure if this will fix #273, but it should help.

### CODE REVIEW

**Comment by Lnaden** [comment]
@jchodera I merged in the changes from #275 to get the tests to run here and hopefully not fail this time.

**Comment by Lnaden** [comment]
Now this is core dumping for some reason. Not failing in the same spot on the test though as the original segfault issue.

I'd like to cut a 0.13.0 release but I wonder if this should be in it first or we bump it to 0.13.1?

**Comment by jchodera** [comment]
@pgrinaway : I just noticed this PR was still open. Do we need to finish/fix this for your changes in `perses` to work?

**Comment by jchodera** [comment]
We seem to have that segfault issue again, so I've restarted the tests:
```
Check that ThermostatedIntegrator copies the correct class and attributes. ... ok (0.0172s)
test_integrators.test_alchemical_langevin_integrator ... /home/travis/.travis/job_stages: line 104:  2849 Segmentation fault      (core dumped) nosetests $PACKAGENAME --nocapture --verbosity=2 --with-timer --with-doctest --doctest-options="+IGNORE_EXCEPTION_DETAIL" -a '!slow'
```

**Comment by jchodera** [comment]
Weird! The tests keep failing with:
```
test_integrators.test_alchemical_langevin_integrator ... /home/travis/.travis/functions: line 104:  5154 Segmentation fault      (core dumped) nosetests $PACKAGENAME --nocapture --verbosity=2 --with-timer --with-doctest -a '!slow'
```

**Comment by jchodera** [comment]
I think the only remaining failure here is the known segfault with `_find_barostat()`.

**Comment by jchodera** [comment]
In debugging the doctests, I've managed to get this short test script to segfault on my local `osx` box:
```python
# Create harmonic oscillator testsystem
from openmmtools import testsystems
testsystem = testsystems.HarmonicOscillator()
# Create a nonequilibrium alchemical integrator
alchemical_functions = { 'testsystems_HarmonicOscillator_x0' : 'lambda' }
nsteps_neq = 100 # number of steps in the switching trajectory where lambda is switched from 0 to 1
from openmmtools import integrators
from simtk import openmm, unit
print('Creating integrator...')
integrator = integrators.AlchemicalNonequilibriumLangevinIntegrator(temperature=300*unit.kelvin, collision_rate=1.0/unit.picoseconds, timestep=1.0*unit.femtoseconds,
                                                                    alchemical_functions=alchemical_functions, splitting="O { V R H R V } O", nsteps_neq=nsteps_neq,
                                                                    measure_shadow_work=True)
```

**Comment by jchodera** [comment]
The local segfault turns out to be due to https://github.com/pandegroup/openmm/issues

File: openmmtools/integrators.py [hunk 1]
```diff
@@ -1774,6 +1777,19 @@ def _step_dispatch_table(self):
...
+        """
+        # Reset statistics
+        super(AlchemicalNonequilibriumLangevinIntegrator, self).reset()
+        # Trigger update of all context parameters only by running one integrator cycle with step = -1
+        self.setGlobalVariableByName('step', -1)
+        self.step(1)
+
+    def _register_step_types(self):
+        super(AlchemicalNonequilibriumLangevinIntegrator, self)._register_step_types()
+        self._register_step_type('H', self._add_alchemical_perturbation_step)
```
[the remark this answers is not in the document]
Comment (jchodera) [reply openmmtools/integrators.py]: Good catch!

**Comment by jchodera** [comment]
Oh no:
```
CompoundThermodynamicState.system and set_system method. ... ok
Test method GlobalParameterState.apply_to_context. ... python(22065,0x7fff8a4bf380) malloc: *** error for object 0x7f88484cf3b8: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
Abort trap: 6
```

**Comment by jchodera** [comment]
CC: https://github.com/pandegroup/openmm/issues/2300

**Comment by jchodera** [comment]
OK, we may want to revert some of these conversions of `try...except` clauses since it seems like they may have just been a buggy environment on my iMac. But I think I did fix at least _one_ segfault that had been happening on travis. We'll see if it's still there in the tests.

**Comment by jchodera** [comment]
Huzzah!

**Comment by jchodera** [comment]
@andrrizzi : For posterity, I think this code gave segfaults durign testing in defining a system with two barostats:
```python
        cls.multiple_barostat_alanine = copy.deepcopy(cls.barostated_alanine)
        barostat = openmm.MonteCarloBarostat(cls.std_pressure, cls.std_temperature)
        cls.multiple_barostat_alanine.addForce(barostat)
```
but this code works fine:
```python
        cls.multiple_barostat_alanine = copy.deepcopy(cls.alanine_explicit)
        barostat = openmm.MonteCarloBarostat(cls.std_pressure, cls.std_temperature)
        cls.multiple_barostat_alanine.addForce(barostat)
        barostat = openmm.MonteCarloBarostat(cls.std_pressure, cls.std_temperature)
        cls.multiple_barostat_alanine.addForce(barostat)
```
That suggests there's some issue with deep copies of `MonteCarloBarostat`.

**Comment by andrrizzi** [comment]
Good to know, thanks! Will keep an eye out if the segfault pops up again.

**Pull request merged**